### PR TITLE
Merit Based Sortition / Permeability of Inferer, Forecaster, Reputer Active Set

### PR DESCRIPTION
**Pull request by relyt29**
## Purpose of Changes and their Description

Due to the scaling constraints of running a blockchain, we can't process all of the inferences, forecasts, and reputation bundles given to us for rewards.

At the time of payload acceptance (both worker and reputer), we only accept up to `MaxTopXToReward` inferences, forecasts, and reputations to process and store on chain. We use a skimming the top algorithm: we accept everything up until we hit the max amount permissible, and then any subsequent payloads replace the payload of the actor with the lowest Exponential Moving Average (EMA) score in the existing set we have already accepted. We do not filter at this point who is allowed to submit payloads other than only "top N by score" accepted.

We also avoid caculating EMA scores for every actor. Because the calculation of scores requires complete knowledge of the network loss for the epoch, EMA score calculation happens at time of reputer nonce closure (in endblock), upon the final set of payloads submitted at that point. 

We order EMA scores into "active" and "passive" sets of inferers, forecasters, and reputers using score EMAs and quantiles.  `FindTopNByScoreDesc` sorts the set of actors based on their scores (roughly, their distance from the combined network loss) and only keeps the `moduleParams.MaxTopXToReward` actors (where `X` is one of inferers, forecasters, or reputers). In order to smooth out the effect of any particular epoch in the scores of actors, we use Exponential Moving Averages of scores with a new alpha parameter `moduleParams.MeritSortitionAlpha` - this replaces the latest scores data structures in the keeper, so `latestXScoresByX` becomes `xScoreEma` in the keeper.

In order to allow for permeability in the active set, i.e. to allow new players without established reputation to get a chance to upload payloads, we take the `topic.ActiveXQuantile` quantile of all the scores (e.g. the 20th percentile), and assign everyone in the passive set the quantile score as their score. This introduces churn in the bottom quantile of the active set, allowing players without established reputation to break into the active set and submit payloads if they are decently good.

Bullet point list of changes follows:

* Added new fields to the Topic struct for active quantiles: topics now include `active_inferer/forecaster/reputer_quantile` parameter between 0 and 1 for permeability.
* Introduced a merit sortition alpha parameter for weighting previous score EMAs:
* Refactored the FindTopNByScoreDesc function to not rely on a heap data structure
* Moved score calculation logic into the actor_utils package and updated the rewards system to get score EMAs from the keeper at time of rewards, rather than calculate scores at time of rewards.
* Renamed util_sort to sortition to better reflect its purpose.
* Moved validations from the types package into the msg server where they are performed.
* Changed the behavior of worker and reputer payload submissions to no longer filter them at the time of submission. Instead, active set management for inferers, forecasters, and reputers is now done at the time of reputer nonce closure.
* Updated proto files to reflect changes in score querying, renaming endpoints from "LatestScore" to "ScoreEma":
* Moved IsBetweenZeroAndOneExclusive function into the math package for broader use.
* Moved types/msg_create_topic.go into the msg_server package.
* Renames the query server tests from KeeperTestSuite to QueryServerTestSuite

Reviewers: The most important changes to focus on in this PR occur in:
* `x/emissions/keeper/actor_utils/losses.go`
* `x/emissions/keeper/msgserver/msg_server_worker_payload.go`
* `x/emissions/keeper/msgserver/msg_server_reputer_payload.go`

## Link(s) to Ticket(s) or Issue(s) resolved by this PR

PROTO-2204, PROTO-1813, RES-297 

This is built upon / replaces #421 

## Are these changes tested and documented?

* Implemented and tested quantile calculations, including CSV-based simulation matching testing.
* Updated keeper tests and added new tests for quantile calculations and score EMAs.

In progress:
- [ ] If tested, please describe how. If not, why tests are not needed.
- [ ] If documented, please describe where. If not, describe why docs are not needed.
- [ ] Added to `Unreleased` section of `CHANGELOG.md`?

## Still Left Todo

* Fix failing CI
* Tests
  * Existing Unit tests to pass
  * New Unit tests to cover new functionality
  * integration tests
* Migrations for upgrade
* Changelog.txt to be updated
* address the todos littered in the PR comments (BAD!)

### CODE REVIEW

File: x/emissions/keeper/actor_utils/sortition.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	ret, err := lowerScore.Score.Add(product)
+	if err != nil {
+		return alloraMath.Dec{}, err
+	}
+	return ret, nil
+}
+
+/*
+func UpdateScoresOfPassiveActorsWithActivePercentile(
```
Comment (kpeluso) [inline x/emissions/keeper/actor_utils/sortition.go:131]: 👀

File: x/emissions/keeper/queryserver/query_server_score.go [hunk 2]
```diff
@@ -6,49 +6,49 @@ import (
 	"github.com/allora-network/allora-chain/x/emissions/types"
 )
 
-func (qs queryServer) GetLatestInfererScore(
+func (qs queryServer) GetInfererScoreEma(
```
Comment (kpeluso) [inline x/emissions/keeper/queryserver/query_server_score.go:9]: @xmariachi take note for impact on forecaster suite when this is merged

File: x/emissions/keeper/msgserver/msg_server_topics.go [hunk 3]
```diff
@@ -5,11 +5,60 @@ import (
...
+	}
+	if !msg.MaxActiveInferersQuantile.IsBetweenZeroAndOneInclusive() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "max active inferers quantile must be between 0 and 1")
+	}
+	if !msg.MaxActiveForecastersQuantile.IsBetweenZeroAndOneInclusive() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "max active forecasters quantile must be between 0 and 1")
+	}
+	if !msg.MaxActiveReputersQuantile.IsBetweenZeroAndOneInclusive() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "max active reputers quantile must be between 0 and 1")
+	}
```
Comment (kpeluso) [inline x/emissions/keeper/msgserver/msg_server_topics.go:55]: I'm worried you put these in the wrong spot... Perhaps I'm remembering something wrong, but I understood these parameters to set bounds on what new topics can set their parameters as (and if these parameters are adjusted down, then the minimum of the existing, immutable topic value and this global is used).

Comment (kpeluso) [reply x/emissions/keeper/msgserver/msg_server_topics.go:55]: I'm actually not sure where the existing params you have are used...

File: x/emissions/keeper/msgserver/msg_server_topics.go [hunk 4]
```diff
@@ -5,11 +5,60 @@ import (
...
 	alloraMath "github.com/allora-network/allora-chain/math"
-	"github.com/allora-network/allora-chain/x/emissions/types"
+	emissionstypes "github.com/allora-network/allora-chain/x/emissions/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
-func (ms msgServer) CreateNewTopic(ctx context.Context, msg *types.MsgCreateNewTopic) (*types.MsgCreateNewTopicResponse, error) {
-	if err := msg.Validate(); err != nil {
+func validateNewTopic(msg *emissionstypes.MsgCreateNewTopic) error {
```
Comment (guilherme-brandao) [inline x/emissions/keeper/msgserver/msg_server_topics.go:13]: If we are adopting this new pattern and moving the validation here, please make sure to move the validations of the other messages as well.

File: x/emissions/keeper/msgserver/msg_server_topics.go (hunk 3 above)
Comment (relyt29) [reply x/emissions/keeper/msgserver/msg_server_topics.go:55]: I don't understand what you're trying to say but if you're talking about existing topics, I haven't written the migrations for this PR yet. We'll have to iterate over existing topics and add this field to them if that is what you are asking

File: x/emissions/proto/emissions/v3/genesis.proto [hunk 5]
```diff
@@ -47,11 +50,11 @@ message GenesisState {
...
   repeated TopicIdBlockHeightScores reputerScoresByBlock = 12;
   // map of (topic, block_height, worker) -> score
-  repeated TopicIdActorIdScore latestInfererScoresByWorker = 13;
+  repeated TopicIdActorIdScore infererScoreEmas = 57;
   // map of (topic, block_height, worker) -> score
-  repeated TopicIdActorIdScore latestForecasterScoresByWorker = 14;
+  repeated TopicIdActorIdScore forecasterScoreEmas = 58;
   // map of (topic, block_height, reputer) -> score
-  repeated TopicIdActorIdScore latestReputerScoresByReputer = 15;
+  repeated TopicIdActorIdScore reputerScoreEmas = 59;
```
Comment (kpeluso) [inline x/emissions/proto/emissions/v3/genesis.proto:57]: should convert to snake case while we have new emissions module version

File: x/emissions/types/allora_type_aliases.go [hunk 6]
```diff
@@ -16,3 +18,11 @@ type TaskReward struct {
 	TopicId TopicId
 	Type    TaskRewardType
 }
+
+// contains a score and a pointer to the
+// actor with the next lowest score,
+// for updating the lowest scores in the keeper
+type ScoreAndNextLowest struct {
+	Score      Score
+	NextLowest string
```
Comment (kpeluso) [inline x/emissions/types/allora_type_aliases.go:27]: There a linked list somewhere I'm missing...?

Comment (relyt29) [reply x/emissions/types/allora_type_aliases.go:27]: Good catch, I started to write a linked list and then decided to not 

This is garbage leftover code

File: x/emissions/keeper/msgserver/msg_server_topics.go (hunk 3 above)
Comment (relyt29) [reply x/emissions/keeper/msgserver/msg_server_topics.go:55]: todo: rename better

**Comment by relyt29** [comment]
closing in favor of #556

**Pull request closed without merging**